freqtrade/freqai/freqai_interface.py

```diff
@@ -0,0 +1,202 @@
+import gc
+import logging
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import Any, Dict, Tuple
+
+import numpy.typing as npt
+import pandas as pd
+from pandas import DataFrame
+
+from freqtrade.enums import RunMode
+from freqtrade.freqai.data_kitchen import FreqaiDataKitchen
+from freqtrade.strategy.interface import IStrategy
+
+
+pd.options.mode.chained_assignment = None
+logger = logging.getLogger(__name__)
+
+
+class IFreqaiModel(ABC):
+    """
+    Class containing all tools for training and prediction in the strategy.
+    User models should inherit from this class as shown in
+    templates/ExamplePredictionModel.py where the user overrides
+    train(), predict(), fit(), and make_labels().
+    Author: Robert Caulk, rob.caulk@gmail.com
+    """
+
+    def __init__(self, config: Dict[str, Any]) -> None:
+
+        self.config = config
+        self.freqai_info = config["freqai"]
+        self.data_split_parameters = config["freqai"]["data_split_parameters"]
+        self.model_training_parameters = config["freqai"]["model_training_parameters"]
+        self.feature_parameters = config["freqai"]["feature_parameters"]
+        # self.backtest_timerange = config["timerange"]
+
+        self.time_last_trained = None
+        self.current_time = None
+        self.model = None
+        self.predictions = None
+
+    def start(self, dataframe: DataFrame, metadata: dict, strategy: IStrategy) -> DataFrame:
+        """
```

Comment:
Do i understand this right that this will either predict - or train - based on the runmode (for now)?
How would you imagine this to work with automatic retraining?

I think this should be splitup to a method doing "train" and "predict()" - assuming the training in live will also use the sliding window approach (??)

Reply:
No, I should correct your understanding:

`start()` will train a new model if it cannot find a pretrained model. **Live** and **backtesting**.
if `start()` does find find a pretrained model, it will simply load that model and inference (use predict()) it. **Live** and **backtesting**

The only difference for **live** is that if it finds an existing model, it checks the amount of time that has elapsed since the end of that training period. If the elapsed time is greater than the config user defined `backtest_period`, then it trains a new model and points any future inferences (predict()) to the new model. 

So your understanding that backtesting = training and live = predicting is incorrect. We train and predict in both backtesting and live. In other words, we need `start()` to detect if we need training/predicting in both backtesting and live. 

Please let me know if it makes sense, and if not, let me know if I can clarify things more.

Reply:
thanks - it "kindof" makes sense now - except that it doesn't - or rather - it raises more questions.

In my understanding, the goal is to move the "live retraining" to a separate thread/process - to avoid impacts on "live trades".

By calling start (and training there) from this part of the strategy - we're effectively retraining "in-line" (or in-process) - slowing "this iteration" down by "training + prediction" (however long this is).

this might not be a problem - but depending on the trade state (close to a stoploss, maybe) - we might be missing stops/exits while the model is being retrained as the bot-loop is blocked - which is clearly not a safe way for this.

Maybe this can set a flag for the retraining (this is where the "non-repeated initialization" comes into play) - which will let freqtrade know that "now we need to retrain everything, please spawn new processes for me".
Obviously there's alternative approaches (like a "diff-thread-callback") which is used for this "out of process" processing (that's not the point of this comment) - but my point is - having this in-line in live (!) is probably not the best idea

Reply:
>By calling start (and training there) from this part of the strategy - we're effectively retraining "in-line" (or in-process) - slowing "this iteration" down by "training + prediction" (however long this is).
>having this in-line in live (!) is probably not the best idea

You are 100% correct :-) And moving the training to a separate thread is on the road map - I will do that after I sort out the parameters.